File: Two_dimensional_array/java/twoDArray.java
public class twoDArray {

    // Method to calculate the sum of all elements in a 2D array
    public static int sumAll(int[][] mat) {
        int r = mat.length;             // Number of rows
        int c = mat[0].length;          // Number of columns
        int sum = 0;                    // Initialize sum

        // Iterate through each element in the matrix and add to sum
        for (int i = 0; i < r; i++) {
            for (int j = 0; j < c; j++) {
                sum += mat[i][j];
            }
        }

        return sum;                     // Return the calculated sum
    }

    // Method to calculate column-wise sum and print the result
    public static void columnSum(int[][] mat) {
        int row = mat.length;           // Number of rows
        int col = mat[0].length;        // Number of columns
        int[] arr = new int[col];       // Array to store column sums

        // Calculate column sums
        for (int j = 0; j < col; j++) { // Loop through each column
            for (int i = 0; i < row; i++) { // Loop through each row of the current column
                arr[j] += mat[i][j];    // Add current element to column sum
            }
        }

        // Print each column sum side by side
        for (int i = 0; i < arr.length; i++) {
            System.out.print(arr[i] + " ");
        }
        System.out.println();           // Move to the next line after printing sums
    }

    // Method to calculate row-wise sum and print the result matrix
    public static void rowSum(int[][] mat) {
        int r = mat.length;             // Number of rows
        int c = mat[0].length;          // Number of columns
        int[][] resultedArray = new int[r][1]; // Resultant array for row sums

        // Calculate row sums
        for (int i = 0; i < r; i++) {   // Loop through each row
            for (int j = 0; j < c; j++) { // Loop through each column of the current row
                resultedArray[i][0] += mat[i][j]; // Add current element to row sum
            }
        }

        printMatrix(resultedArray);     // Print the resultant matrix
    }

    // Method to multiply two matrices and print the result matrix
    public static void multiplyMatrix(int[][] matA, int[][] matB) {
        int ar = matA.length;           // Rows in matrix A
        int ac = matA[0].length;        // Columns in matrix A
        int br = matB.length;           // Rows in matrix B
        int bc = matB[0].length;        // Columns in matrix B

        int[][] result = new int[ar][bc]; // Resultant matrix

        // Matrix multiplication logic
        for (int i = 0; i < ar; i++) {  // Loop through rows of matrix A
            for (int j = 0; j < bc; j++) { // Loop through columns of matrix B
                for (int k = 0; k < br; k++) { // Loop through rows of matrix B
                    result[i][j] += matA[i][k] * matB[k][j]; // Perform multiplication and sum
                }
            }
        }

        printMatrix(result);            // Print the resultant matrix
    }

    

    // Method to print the diagonal elements of a matrix
    public static void printDiagonal(int[][] mat) {
        for (int i = 0; i < mat.length; i++) {
            System.out.println(mat[i][i]); // Print diagonal element
        }
    }

    // Method to print the second diagonal elements of a matrix
    public static void printSecondDiagonal(int[][] mat) {
        int col = mat.length;           // Number of columns

        // Print second diagonal elements
        for (int i = 0; i < mat.length; i++) {
            System.out.println(mat[i][col - 1 - i]);
        }
    }

    // Here in printMatrix function multiple inbuilt functions are used to make the formatting look good. You are not allowed to use this in your code.
    public static void printMatrix(int[][] mat) {
        int r = mat.length;
        int c = mat[0].length;

        // Find the maximum length of each column
        int[] max_lengths = new int[c];
        for (int j = 0; j < c; j++) {
            max_lengths[j] = Integer.MIN_VALUE;
            for (int i = 0; i < r; i++) {
                int length = String.valueOf(mat[i][j]).length();
                if (length > max_lengths[j]) {
                    max_lengths[j] = length;
                }
            }
        }

        // Print top border
        printHorizontalLine(max_lengths);

        // Print matrix with formatted cells
        for (int i = 0; i < r; i++) {
            StringBuilder formattedRow = new StringBuilder("| ");
            for (int j = 0; j < c; j++) {
                formattedRow.append(String.format("%" + max_lengths[j] + "d", mat[i][j]));
                if (j < c - 1) {
                    formattedRow.append(" | ");
                }
            }
            formattedRow.append(" |");

            System.out.println(formattedRow);

            // Print inner horizontal line between rows, except after the last row
            if (i < r - 1) {
                printHorizontalLine(max_lengths);
            }
        }

        // Print bottom border
        printHorizontalLine(max_lengths);
    }

    // Helper method to print horizontal line
    private static void printHorizontalLine(int[] max_lengths) {
        StringBuilder line = new StringBuilder("+");
        for (int length : max_lengths) {
            line.append("-".repeat(length + 2)).append("+");
        }
        System.out.println(line);
    }
    // Main method for testing purposes
    public static void main(String[] args) {
        int[][] mat2 = { { 1, 2, 3 }, { 44, 555, 66 }, { 7, 8, 9 } };

        // Test cases
        System.out.println("Original Matrix:");
        printMatrix(mat2);

        System.out.println("\nColumn Sum:");
        columnSum(mat2);

        System.out.println("\nDiagonal Elements:");
        printDiagonal(mat2);

        System.out.println("\nSecond Diagonal Elements:");
        printSecondDiagonal(mat2);

        int[][] mat3 = { { 1, 2 }, { 3, 4 }, { 5, 6 } };

        System.out.println("\nMultiply Matrix:");
        multiplyMatrix(mat2, mat3);

        System.out.println("\nRow Sum:");
        rowSum(mat2);

        System.out.println("\nSum of all elements: " + sumAll(mat2));
    }
}

File: Two_dimensional_array/python/twoDArray.py
import numpy as np

# Here in printMatrix function multiple inbuilt functions are used to make the formatting look good. You are not allowed to use this in your code.

def printMatrix(mat):
    """
    Prints a matrix with formatted output.
    """
    r, c = mat.shape
    max_lengths = [max(len(str(mat[i, j])) for i in range(r)) for j in range(c)]
    rowChars = []
    for i in range(r):
        formatted_row = '| ' + ' | '.join(f"{mat[i, j]:>{max_lengths[j]}}" for j in range(c)) + ' | '
        rowChars.append(formatted_row)
    for i in range(len(rowChars)):
        print('-'*(len(rowChars[i])-1))
        print(rowChars[i])
        if i == len(rowChars)-1:
            print('-'*(len(rowChars[i])-1))

def columnSum(matrix1):
    """
    Calculates and prints the sum of each column in a matrix.
    """
    r, c = matrix1.shape
    resultedArray = np.zeros(c, dtype=int)
    for j in range(c):
        for i in range(r):
            resultedArray[j] += matrix1[i][j]
    print(resultedArray)

def printDiagonal(matrix1):
    """
    Prints the elements of the main diagonal of a square matrix.
    """
    r, c = matrix1.shape
    for i in range(r):
        print(matrix1[i][i])

def printSecondDiagonal(mat):
    """
    Prints the elements of the secondary diagonal of a square matrix.
    """
    r, c = mat.shape
    for i in range(r):
        print(mat[i][r-1-i])


def sumAll(mat):
    """
    Calculates the sum of all elements in a 2D array.
    """
    r, c = mat.shape # Get the number of rows and columns
    total_sum = 0 # Initialize the sum
    for i in range(r): # Iterate through each row
        for j in range(c): # Iterate through each column
            total_sum += mat[i][j] # Add the element to the sum
    return total_sum # Return the total sum

def rowSum(mat):
    """
    Calculates and prints the row sums of a matrix.
    """
    r, c = mat.shape # Get the number of rows and columns
    resultedArray = np.zeros((r, 1), dtype=int) # Initialize the result array
    for i in range(r): # Iterate through rows
        for j in range(c): # Iterate through columns
            resultedArray[i][0] += mat[i][j] # Calculate the row sum
    printMatrix(resultedArray) # Print the row sums

def multiplyMatrix(matA, matB):
    """
    Multiplies two matrices and prints the result.
    """
    ar, ac = matA.shape # Dimensions of matrix A
    br, bc = matB.shape # Dimensions of matrix B
    result = np.zeros((ar, bc), dtype=int) # Initialize the result matrix
    for i in range(ar): # Iterate through rows of A
        for j in range(bc): # Iterate through columns of B
            for k in range(br): # Iterate through rows of B
                result[i][j] += matA[i][k] * matB[k][j] # Perform matrix multiplication
    printMatrix(result) # Print the result matrix

def print_spiral(matrix):
    """
    Print matrix in spiral order.
    """
    # Check for empty matrix
    if not matrix or not matrix[0]:
        return
    # Define matrix boundaries
    row_start, row_end = 0, len(matrix) - 1
    col_start, col_end = 0, len(matrix[0]) - 1
    # Traverse the matrix in spiral order
    while row_start <= row_end and col_start <= col_end:
        # Traverse top row (left → right)
        for c in range(col_start, col_end + 1):
            print(matrix[row_start][c], end=" ")
        row_start += 1  # Move top boundary down
        # Traverse right column (top → bottom)
        for r in range(row_start, row_end + 1):
            print(matrix[r][col_end], end=" ")
        col_end -= 1  # Move right boundary left
        # Traverse bottom row (right → left), if rows remain
        if row_start <= row_end:
            for c in range(col_end, col_start - 1, -1):
                print(matrix[row_end][c], end=" ")
            row_end -= 1  # Move bottom boundary up
        # Traverse left column (bottom → top), if columns remain
        if col_start <= col_end:
            for r in range(row_end, row_start - 1, -1):
                print(matrix[r][col_start], end=" ")
            col_start += 1  # Move left boundary right
    print()  # Newline after full traversal

# Main method with test cases
if __name__ == "__main__":
    # Test case for printMatrix
    mat_test = np.array([[1, 2, 3], [4, 5, 6], [7, 8, 9]])
    print("Test Case printMatrix:")
    printMatrix(mat_test)

    # Test case for columnSum
    mat_sum_test = np.array([[1, 2, 3], [4, 5, 6]])
    print("\nTest Case columnSum:")
    columnSum(mat_sum_test)

    # Test case for printDiagonal
    mat_diag_test = np.array([[1, 2, 3], [4, 5, 6], [7, 8, 9]])
    print("\nTest Case printDiagonal:")
    printDiagonal(mat_diag_test)

    # Test case for printSecondDiagonal
    mat_sec_diag_test = np.array([[1, 2, 3], [4, 5, 6], [7, 8, 9]])
    print("\nTest Case printSecondDiagonal:")
    printSecondDiagonal(mat_sec_diag_test)


    # Test case for sumAll
    mat_sum_all_test = np.array([[1, 2, 3], [4, 5, 6]])
    print("\nTest Case sumAll:")
    print(sumAll(mat_sum_all_test))

    # Test case for rowSum
    mat_row_sum_test = np.array([[1, 2, 3], [4, 5, 6], [7, 8, 9]])
    print("\nTest Case rowSum:")
    rowSum(mat_row_sum_test)

    # Test case for multiplyMatrix
    mat_multiply_test_A = np.array([[1, 2], [3, 4], [5, 6]])
    mat_multiply_test_B = np.array([[5, 6], [7, 8]])
    print("\nTest Case multiplyMatrix:")
    multiplyMatrix(mat_multiply_test_A, mat_multiply_test_B)
